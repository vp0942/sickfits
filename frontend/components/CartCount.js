import styled from 'styled-components';
import PropTypes from 'prop-types';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

const Dot = styled.div`
  background: var(--red);
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-feature-settings: 'tnum'; // This is to make sure that the font size doesn't change when the number of items in the cart changes
  font-variant-numeric: tabular-nums; // This is to make sure that the font size doesn't change when the number of items in the cart changes
`;

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: transform 0.4s;
    backface-visibility: hidden; // This is to prevent the text from being visible from the back when it's being animated
  }
  /* Initial state of the entered Dot */
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  /* Intermediate state of the entered Dot */
  .count-enter-active {
    transform: rotateX(0);
  }
  /* Final state of the entered Dot */
  .count-exit {
    top: 0; // This is to prevent the Dot from moving down when it's being animated
    position: absolute;
    transform: rotateX(0);
  }
  /* Intermediate state of the exited Dot */
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}

CartCount.propTypes = {
  count: PropTypes.number.isRequired,
};
