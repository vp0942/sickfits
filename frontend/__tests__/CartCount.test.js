import { render, screen } from '@testing-library/react';
import CartCount from '../components/CartCount';
import wait from 'waait';

describe('<CartCount />', () => {
  it('Renders', () => {
    render(<CartCount count={10} />)
  });

  it('Matches the snapshot', () => {
    const { container } = render(<CartCount count={11} />);
    expect(container).toMatchSnapshot();
  });

  it('updates via props', async () => {
    const { container, rerender, debug } = render(<CartCount count={11} />);
    expect(container.textContent).toBe('11');
    // expect(container).toHaveTextContent('11'); // Exactly the same as the above line

    rerender(<CartCount count={12} />);
    expect(container.textContent).toBe('1211');
    // We need to wait for the animation to finish - otherwise, the test will fail
    await wait(400);
    expect(container.textContent).toBe('12');
    // debug();
    expect(container).toMatchSnapshot();

  });

});