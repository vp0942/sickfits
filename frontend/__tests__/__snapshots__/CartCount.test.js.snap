// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount /> Matches the snapshot 1`] = `
<div>
  <span
    class="CartCount__AnimationStyles-sc-xxvp4g-1 kHAPbE"
  >
    <div>
      <div
        class="CartCount__Dot-sc-xxvp4g-0 gRrnjj count"
      >
        11
      </div>
    </div>
  </span>
</div>
`;

exports[`<CartCount /> updates via props 1`] = `
<div>
  <span
    class="CartCount__AnimationStyles-sc-xxvp4g-1 kHAPbE"
  >
    <div>
      <div
        class="CartCount__Dot-sc-xxvp4g-0 gRrnjj count count-enter-done"
      >
        12
      </div>
    </div>
  </span>
</div>
`;
